### PR TITLE
Features tests

### PR DESCRIPTION
**Pull request by LLambertRGi**
This change is [Reviewable](https://reviewable.io/reviews/githubrgi/swagd/382)

### CODE REVIEW

File: GeoPackage/src/main/java/com/rgi/geopackage/features/AbstractColumnDefinition.java [hunk 1]
```diff
@@ -116,7 +118,7 @@ protected AbstractColumnDefinition(final String              name,
         this.defaultValue    = defaultValue;
         this.comment         = comment;
 
-        if(this.comment.contains("\n"))
+        if(this.comment != null && this.comment.contains("\n"))
         {
             throw new IllegalArgumentException("The comment my not contain any line break characters");
```
Comment (JACochran) [inline GeoPackage/src/main/java/com/rgi/geopackage/features/AbstractColumnDefinition.java]: my should be may*

File: GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryHeader.java [hunk 2]
```diff
@@ -134,20 +157,21 @@ protected BinaryHeader(final byte                      version,
...
+        @SuppressWarnings("NumericCastThatLosesPrecision")
         final int envelopeContentsMask = (byte)(this.envelopeContentsIndicator.getCode() << 1);
 
-        this.flags = // 0                         |
-                     this.binaryType.getBitMask() |
-                     isEmptyMask                  |
-                     envelopeContentsMask         |
-                     (byteOrder.equals(ByteOrder.BIG_ENDIAN) ? 0 : 1);
+        //noinspection NumericCastThatLosesPrecision
+        this.flags = (byte)(// 0                         |
```
Comment (JACochran) [inline GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryHeader.java]: is there a reason for 0 to be commented out?

File: GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryHeader.java [hunk 3]
```diff
@@ -156,47 +180,142 @@ protected BinaryHeader(final byte                      version,
...
-            byteBuffer.putDouble(envelopeBound);
+            byteOutputStream.write(envelopeBound);
         }
     }
 
-    public static void writeBytes(final ByteBuffer byteBuffer,
-                                  final Geometry   geometry,
-                                  final int        spatialReferenceSystemIdentifier) throws IOException
+    /**
+     * Constructors a header from the supplied arguments, and writes it to a
```
Comment (JACochran) [inline GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryHeader.java]: Should Constructors be Constructs?

**Comment by JACochran** [comment]
Reviewed 82 of 167 files at r1, 11 of 17 files at r2.
Review status: 93 of 183 files reviewed at latest revision, 4 unresolved discussions.

---

*[GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryHeader.java, line 114 \[r3\]](https://reviewable.io:443/reviews/githubrgi/swagd/382#-KH1IIqzFGjjBrKkHmFy:-KH1IIr-jVcFPPkh70HZ:1514021626) ([raw file](https://github.com/githubrgi/swagd/blob/50fbd8cb6cda8fd82629c1c0b6433ab17ac9f129/GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryHeader.java#L114)):*
envelop's should be envelope's*

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/githubrgi/swagd/382)*

**Comment by JACochran** [comment]
Reviewed 1 of 167 files at r1.
Review status: 94 of 183 files reviewed at latest revision, 5 unresolved discussions.

---

*[GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryType.java, line 39 \[r3\]](https://reviewable.io:443/reviews/githubrgi/swagd/382#-KH1JwjGWBMyTtvL757k:-KH1JwjK-OXXP10v2nJp:-2017205569) ([raw file](https://github.com/githubrgi/swagd/blob/50fbd8cb6cda8fd82629c1c0b6433ab17ac9f129/GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryType.java#L39)):*
remove todo.  Not obvious what these are, please add comment :)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/githubrgi/swagd/382)*

**Comment by JACochran** [comment]
Reviewed 1 of 167 files at r1.
Review status: 20 of 117 files reviewed at latest revision, 6 unresolved discussions.

---

*[GeoPackage/src/main/java/com/rgi/geopackage/features/ByteOutputStream.java, line 32 \[r3\]](https://reviewable.io:443/reviews/githubrgi/swagd/382#-KH1LonvDkzQ0VrUyKsN:-KH1LonxDXFEuaprxz_5:160151489) ([raw file](https://github.com/githubrgi/swagd/blob/50fbd8cb6cda8fd82629c1c0b6433ab17ac9f129/GeoPackage/src/main/java/com/rgi/geopackage/features/ByteOutputStream.java#L32)):*
Not many comments in this class including on public methods.  It would be a good idea to add some though most of it is obvious.  Just a suggestion

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/githubrgi/swagd/382)*

**Comment by JACochran** [comment]
Reviewed 6 of 167 files at r1.
Review status: 26 of 117 files reviewed at latest revision, 8 unresolved discussions.

---

*[GeoPackage/src/main/java/com/rgi/geopackage/features/FeatureSet.java, line 92 \[r3\]](https://reviewable.io:443/reviews/githubrgi/swagd/382#-KH1O7wxK-CfXMDs-w12:-KH1O7wzMwzKiO9n_ld6:974504337) ([raw file](https://github.com/githubrgi/swagd/blob/50fbd8cb6cda8fd82629c1c0b6433ab17ac9f129/GeoPackage/src/main/java/com/rgi/geopackage/features/FeatureSet.java#L92)):*
may not be null OR empty*

---

*[GeoPackage/src/main/java/com/rgi/geopackage/features/FeatureSet.java, line 97 \[r3\]](https://reviewable.io:443/reviews/githubrgi/swagd/382#-KH1OC7PXDb_P3oC5-KI:-KH1OC7PXDb_P3oC5-KJ:144863385) ([raw file](https://github.com/githubrgi/swagd/blob/50fbd8cb6cda8fd82629c1c0b6433ab17ac9f129/GeoPackage/src/main/java/com/rgi/geopackage/features/FeatureSet.java#L97)):*
may not be null OR empty*

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/githubrgi/swagd/382)*

**Comment by JACochran** [comment]
Reviewed 74 of 167 files at r1, 5 of 17 files at r2, 3 of 3 files at r3, 9 of 79 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions.

---

*[GeoPackage/src/main/java/com/rgi/geopackage/features/GeometryType.java, line 32 \[r3\]](https://reviewable.io:443/reviews/githubrgi/swagd/382#-KH1P35ykoKNWFyU1jZs:-KH1P362n6vvQGVJ9lS_:567759719) ([raw file](https://github.com/githubrgi/swagd/blob/50fbd8cb6cda8fd82629c1c0b6433ab17ac9f129/GeoPackage/src/main/java/com/rgi/geopackage/features/GeometryType.java#L32)):*
did you mean to leave this here?

---

*[GeoPackage/src/main/java/com/rgi/geopackage/features/GeoPackageFeatures.java, line 868 \[r4\]](https://reviewable.io:443/reviews/githubrgi/swagd/382#-KH1TaW8mM9IYAGVJp-r:-KH1TaW8mM9IYAGVJp-s:102553258) ([raw file](https://github.com/githubrgi/swagd/blob/9d8e2da0de60cb6a5eb035948800ff5447231791/GeoPackage/src/main/java/com/rgi/geopackage/features/GeoPackageFeatures.java#L868)):*
maybe mention if the range is inclusive or not inclusive

---

*[GeoPackage/src/test/java/com/rgi/geopackage/features/SqlTypeTest.java, line 49 \[r4\]](https://reviewable.io:443/reviews/githubrgi/swagd/382#-KH1_1AcdRRy_fuclr76:-KH1_1AdDnM5q-RNZ9BF:1479621304) ([raw file](https://github.com/githubrgi/swagd/blob/9d8e2da0de60cb6a5eb035948800ff5447231791/GeoPackage/src/test/java/com/rgi/geopackage/features/SqlTypeTest.java#L49)):*
Can't this class be ignored due to it being an enum? So that it won't effect the coverage

---

*[GeoPackage/src/test/java/com/rgi/geopackage/features/ValueRequirementTest.java, line 78 \[r4\]](https://reviewable.io:443/reviews/githubrgi/swagd/382#-KH1_qdWRVBowlZkAdEF:-KH1_qdXsmYsDHIijO4L:1760401770) ([raw file](https://github.com/githubrgi/swagd/blob/9d8e2da0de60cb6a5eb035948800ff5447231791/GeoPackage/src/test/java/com/rgi/geopackage/features/ValueRequirementTest.java#L78)):*
instead of catching the IAE  have (expected = IllegalArgumentException.class) next to the annotation(@Test) this will make the junit test fail if this exception is not thrown.  Then this can be split into different tests.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/githubrgi/swagd/382)*

File: GeoPackage/src/main/java/com/rgi/geopackage/features/AbstractColumnDefinition.java (hunk 1 above)
Comment (LLambertRGi) [inline GeoPackage/src/main/java/com/rgi/geopackage/features/AbstractColumnDefinition.java]: fixed

File: GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryHeader.java (hunk 2 above)
Comment (LLambertRGi) [inline GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryHeader.java]: Bitwise or with a 0 is an effective NOOP, which intellij's inspections rightly complain about. It was originally there to indicate that the first "reserved" (according to the standard) bitsshould start out as 0. They'll be 0 either way, it was just a visual indication. I'll take that out.

File: GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryHeader.java (hunk 3 above)
Comment (LLambertRGi) [inline GeoPackage/src/main/java/com/rgi/geopackage/features/BinaryHeader.java]: fixed

**Pull request merged**